### PR TITLE
A0-1139: Clean up e2e tests

### PR DESCRIPTION
**Pull request by maciejzelaszczyk**
We would like to clean up e2e tests:

-  We extract common code that reads storage to `aleph_client`, e.g. getting `CommitteeSize` in `check_points`.
- We unify the naming for these functions, e.g. `get_*, following the existing `aleph_client` convention.
- We extract common parts of tests into helper functions, e.g. for node setup, config parsing, etc.

## Type of change

- Refactor (non-breaking change which does not add functionality)

# Checklist:

- I have made corresponding changes to the existing documentation

### CODE REVIEW

File: aleph-client/src/lib.rs [hunk 1]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+        .collect()
+}
+
+#[derive(Decode)]
+pub struct EraValidators {
+    pub reserved: Vec<AccountId>,
+    pub non_reserved: Vec<AccountId>,
+}
+
+pub fn get_current_era_validators(connection: &SignedConnection) -> anyhow::Result<Vec<AccountId>> {
```
Comment (kostekIV) [inline aleph-client/src/lib.rs:445]: this and the following functions return `Result` but they always return the `Ok` option. consider just returning without the wrapping it in the `Result`

File: aleph-client/src/session.rs [hunk 2]
```diff
@@ -137,3 +137,23 @@ pub fn wait_for_at_least<C: AnyConnection>(
...
+    session: u32,
+    session_period: u32,
+) -> Vec<AccountId> {
+    let first_block = session_period * session;
+
+    let block = connection
+        .as_connection()
+        .get_block_hash(Some(first_block))
+        .expect("Api call should succeed")
+        .expect("Session already started so the first block should be present");
```
Comment (kostekIV) [inline aleph-client/src/session.rs:152]: we do not know this here in this function whether its true

File: e2e-tests/src/test/validators_rotate.rs [hunk 3]
```diff
@@ -1,78 +1,29 @@
...
-    let connection = SignedConnection::new(node, sender);
-
-    let sudo = get_sudo_key(config);
+    let connection = get_signed_connection(config);
+    let root_connection = config.create_root_connection();
 
-    let root_connection = RootConnection::new(node, sudo);
     let reserved_validators: Vec<_> = get_reserved_validators(config)
         .iter()
         .map(|pair| AccountId::from(pair.public()))
```
Comment (pmikolajczyk41) [inline e2e-tests/src/test/validators_rotate.rs:29]: this mapping seeds to AccountId seems a common thing, so maybe we can provide some utility in `e2e-tests/src/accounts.rs` (there are already similar methods there)

Comment (pmikolajczyk41) [reply e2e-tests/src/test/validators_rotate.rs:29]: `e2e-tests/src/test/era_validators.rs` would benefit from that as well

**Review by pmikolajczyk41 — COMMENTED** [review]
basically, this PR is pure gold, and I'm really happy that we are doing this! kudos!

I have left some minor comments and recommendations - I'm aware that most of the things I was commenting have been already present before these changes, however, I believe that we can change them right away

also, you may want to check also `e2e-tests/src/test/electing_validators.rs`, since there are some things that could be changed along - e.g. moving `chill()`to `aleph-client`

File: aleph-client/src/lib.rs (hunk 1 above)
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Good catch, fixed.

File: aleph-client/src/session.rs (hunk 2 above)
Comment (maciejzelaszczyk) [reply aleph-client/src/session.rs]: Fixed the message.

File: aleph-client/src/elections.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    connection: &C,
+    block_hash: Option<H256>,
+) -> CommitteeSeats {
+    connection
+        .as_connection()
+        .get_storage_value("Elections", "CommitteeSize", block_hash)
+        .expect("Failed to decode CommitteeSize extrinsic!")
+        .unwrap_or_else(|| {
+            panic!(
+                "Failed to obtain CommitteeSize for block hash: {:?}.",
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/elections.rs]: Yes, I think you are right, changed the ordering.

File: aleph-client/src/elections.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        .expect("Failed to decode SessionValidatorBlockCount extrinsic!")
+}
+
+#[derive(Decode)]
+pub struct EraValidators {
+    pub reserved: Vec<AccountId>,
+    pub non_reserved: Vec<AccountId>,
+}
+
+pub fn get_current_era_validators(connection: &SignedConnection) -> Vec<AccountId> {
```
Comment (fixxxedpoint) [inline aleph-client/src/elections.rs:43]: You are not using it anywhere.

File: aleph-client/src/elections.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+pub fn get_validator_block_count<C: AnyConnection>(
+    connection: &C,
+    account_id: &AccountId,
+    block_hash: Option<H256>,
+) -> Option<u32> {
+    connection
+        .as_connection()
+        .get_storage_map(
+            "Elections",
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/elections.rs]: Wonderful, done it for other mods as well.

File: aleph-client/src/lib.rs [hunk 7]
```diff
@@ -82,6 +85,7 @@ impl FromStr for WsRpcClient {
 }
 
 pub type BlockNumber = u32;
+pub type Hash = H256;
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: None whatsoever, removed.

File: aleph-client/src/lib.rs [hunk 8]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+pub fn get_current_block_number<C: AnyConnection>(connection: &C) -> u32 {
+    connection
+        .as_connection()
+        .get_header::<Header>(None)
+        .expect("Could not fetch header")
+        .expect("Block exists; qed")
+        .number
+}
+
+pub fn get_stakers_as_storage_keys<C: AnyConnection>(
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Agreed. I had some name-wrangling with the mirror function for manual extraction based on a `StorageKey`. I extracted it to this mod, but the name `get_stakers_as_storage_keys_from_storage_key` is rather ugly. If you have any suggestions, I am all ears :)

File: aleph-client/src/session.rs [hunk 9]
```diff
@@ -137,3 +138,22 @@ pub fn wait_for_at_least<C: AnyConnection>(
 pub fn get_session_period<C: AnyConnection>(connection: &C) -> u32 {
     connection.read_constant("Elections", "SessionPeriod")
 }
+
+pub fn get_authorities_for_session<C: AnyConnection>(
+    connection: &C,
+    session: u32,
+    session_period: u32,
```
Comment (fixxxedpoint) [inline aleph-client/src/session.rs:143]: Suggestion: I would remove this parameter.

File: aleph-client/src/session.rs [hunk 10]
```diff
@@ -137,3 +138,22 @@ pub fn wait_for_at_least<C: AnyConnection>(
...
+    let block = get_block_hash(connection, first_block);
+
+    connection
+        .as_connection()
+        .get_storage_value("Session", "Validators", Some(block))
+        .expect("Failed to decode Validators extrinsic!")
+        .expect("Authorities should always be present")
+}
+
+pub fn get_current_validators<C: AnyConnection>(connection: &C) -> Vec<AccountId> {
```
Comment (fixxxedpoint) [inline aleph-client/src/session.rs:155]: It's not consistently named compared with previous method: something like `get_current_authorities`.

File: e2e-tests/src/rewards.rs [hunk 11]
```diff
@@ -239,3 +233,54 @@ pub fn check_points(
...
+                &root_connection.as_signed(),
+                &validator_id,
+                *additional_stake + TOKEN,
+                XtStatus::Finalized,
+            );
+            let stash_connection = SignedConnection::new(node, account_keys.validator.clone());
+            let xt = stash_connection
+                .as_connection()
+                .staking_bond_extra(*additional_stake);
+            send_xt(
```
Comment (fixxxedpoint) [inline e2e-tests/src/rewards.rs:278]: Move this to `alieph-client`.

**Review by fixxxedpoint — CHANGES_REQUESTED** [review]
Good work! Left few small comments.

File: aleph-client/src/lib.rs [hunk 12]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+        .map(|acc| {
+            connection
+                .as_connection()
+                .metadata
+                .storage_double_map_key("Staking", "ErasStakers", era, acc)
+                .unwrap()
+        })
+        .collect()
+}
+
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Yup, went away and moved it.

File: aleph-client/src/lib.rs [hunk 13]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+#[derive(Decode)]
+pub struct EraValidators {
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Yes, we can, changed it.

File: aleph-client/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Ahh, `AnyConnection` will do. Changed.

File: aleph-client/src/lib.rs [hunk 14]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+    pub reserved: Vec<AccountId>,
+    pub non_reserved: Vec<AccountId>,
+}
+
+pub fn get_current_era_validators(connection: &SignedConnection) -> anyhow::Result<Vec<AccountId>> {
+    let eras_validators: EraValidators =
+        connection.read_storage_value("Elections", "CurrentEraValidators");
+    let eras_validators: Vec<AccountId> = eras_validators
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Problem here is that we are not doing a clean read but we concatenate two intermediate values and we cannot rely on the method return type to get info what to decode to for those intermediate values.

File: aleph-client/src/lib.rs [hunk 15]
```diff
@@ -405,3 +409,78 @@ pub fn get_block_hash<C: AnyConnection>(connection: &C, block_number: BlockNumbe
...
+        .reserved
+        .into_iter()
+        .chain(eras_validators.non_reserved.into_iter())
+        .collect();
+    Ok(eras_validators)
+}
+
+pub fn get_current_era_reserved_validators(
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/lib.rs]: Good idea, changed.

File: aleph-client/src/session.rs [hunk 16]
```diff
@@ -137,3 +137,23 @@ pub fn wait_for_at_least<C: AnyConnection>(
 pub fn get_session_period<C: AnyConnection>(connection: &C) -> u32 {
     connection.read_constant("Elections", "SessionPeriod")
 }
+
+pub fn get_authorities_for_session<C: AnyConnection>(
+    connection: &C,
+    session: u32,
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply aleph-client/src/session.rs]: OK, this seams reasonable, changed.

File: e2e-tests/src/connection.rs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+use aleph_client::SignedConnection;
+
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply e2e-tests/src/connection.rs]: This was probably the thing that was bugging me the most about the whole refactor. Thanks, this is precisely the place I was looking for. Fixed.

File: e2e-tests/src/rewards.rs [hunk 18]
```diff
@@ -239,3 +233,54 @@ pub fn check_points(
         max_relative_difference,
     )
 }
+pub fn get_bench_members(
+    non_reserved_members: Vec<AccountId>,
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply e2e-tests/src/rewards.rs]: I do not find this controversial, other than it being slower in terms of write speed. Changed the two mentioned methods. I would not be opposed to a separate PR, in which we could comb through all the tests and get rid of clone.

File: e2e-tests/src/validators.rs [hunk 19]
```diff
@@ -0,0 +1,36 @@
+use aleph_client::{account_from_keypair, KeyPair};
+use primitives::SessionIndex;
+use substrate_api_client::AccountId;
+
+use crate::{accounts::get_validators_keys, Config};
+
+pub fn get_reserved_validators(config: &Config) -> Vec<KeyPair> {
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply e2e-tests/src/validators.rs:10]: Done.

File: e2e-tests/src/test/validators_change.rs [hunk 20]
```diff
@@ -1,22 +1,17 @@
...
+use crate::{accounts::get_validators_keys, config::Config};
 
 pub fn change_validators(config: &Config) -> anyhow::Result<()> {
     let accounts = get_validators_keys(config);
-    let sudo = get_sudo_key(config);
-
-    let connection = RootConnection::new(&config.node, sudo);
+    let connection = config.create_root_connection();
 
     let reserved_before: Vec<AccountId> =
```
[the remark this answers is not in the document]
Comment (maciejzelaszczyk) [reply e2e-tests/src/test/validators_change.rs:16]: Right, done.

File: e2e-tests/src/test/validators_rotate.rs (hunk 3 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/test/validators_rotate.rs]: Done.

File: aleph-client/src/elections.rs (hunk 5 above)
Comment (maciejzelaszczyk) [reply aleph-client/src/elections.rs]: Fixed as part of the reorganization of the logic between `get_current_era_validators`, `get_current_era_reserved_validators` and `get_current_era_non_reserved_validators`.

File: aleph-client/src/session.rs (hunk 9 above)
Comment (maciejzelaszczyk) [reply aleph-client/src/session.rs:143]: Makes sense, done.

File: aleph-client/src/session.rs (hunk 10 above)
Comment (maciejzelaszczyk) [reply aleph-client/src/session.rs:155]: Right, changed to make it more consistent.

File: e2e-tests/src/rewards.rs (hunk 11 above)
Comment (maciejzelaszczyk) [reply e2e-tests/src/rewards.rs]: Done.

**Review by pmikolajczyk41 — APPROVED** [review]
great work!

**Review by fixxxedpoint — APPROVED** [review]
Great work!

**Pull request merged**